:bug: Flashloans adjustments for Aave Like Protocols

jakubswierczek opened the pull request:
## Ticket URL

[Shortcut](https://app.shortcut.com/oazo-apps/story/11917/frontend-library-issues-with-flashloans)

## Description of Changes
- Return flash loan information with `Multiply Strategy`
- Possibility to force flash loan token for Aave & Spark (Although that will work only on L2s)
- Refactor the Close operation. After depositing a flash loan, we were taking the MAX_UNIT amount of collateral. In the case of short positions (for example, with DAI on Mainnet), that would break the transaction. I have changed the behavior, and now, when the collateral and flash loan assets are the same, I withdraw only the necessary amount for the swap. 

## How to Test

Please try to play around with Aave v3 and Spark on the [preview deployment](https://kh8439-3000.csb.app/). In the order information you will see the flashloan amount. 

## PR Definition of Done

Please ensure the following requirements have been met before marking the PR as ready for review:

- [x] All checks are passing
- [x] PR is linked to a corresponding ticket
- [x] PR title is clear and concise
- [x] Code has been self-reviewed and any fixes or improvements noted (See Code review standards in Notion)
- [x] Documentation has been updated if necessary

jakubswierczek committed:
Add flashloan data to strategy

Introduced a new data structure, WithFlashloan, in types/strategies.ts. This structure is to hold flashloan data. The WithFlashloan data type is now used in where the IStrategy simulation object is used. Refreshed other packages to match new structure and types. This change is made in order to have a clear and separated identification and handling of flashloan data across the entire application procedure.

jakubswierczek committed:
Refactor to better handle flashloan data in strategies

Adjusted the data structure to efficiently handle the flashloan data in our strategies. Created a new type, WithFlashloan, which now holds the information regarding flashloan. Updated all the packages where this data type is used. This refactor adds clarity and separation for handling flashloan data during entire application workflow.

jakubswierczek committed:
Updated flashloan data handling in strategies

Modified flashloan data structure in the strategies for improved handling and clarity. A new type, WithFlashloan, has been introduced, which carries all the necessary flashloan information. Made necessary updates in all packages to align with this change. This modification provides a clearer and more isolated handling of flashloan data throughout the application workflow.

jakubswierczek committed:
Fix: Ensure accurate withdrawal amount in AAVE close operation

Fixed a bug in AAVE v3 close operation where the maximum uint value was used causing potential problems with very large withdrawals. Replaced MAX_UINT with the actual collateralAmountToBeSwapped to accurately reflect the amount intended to be withdrawn from AAVE, thus ensuring accurate operation of the close action in the AAVE strategy.

jakubswierczek committed:
Update withdrawal amount in AAVE v3 close operation

Modified the withdrawal amount in AAVE v3 close operation to emit the given collateral amount instead of the previous maximum uint value, which may cause issues for larger withdrawals. This is essential to accurately relay the amount that will be withdrawn and to ensure precise operation of the AAVE strategy.

jakubswierczek committed:
Add symbol to Aave token in risk increasing condition

Updated dma-library to include the debtToken symbol in AAVE strategy while building operation in both risk increasing and risk decreasing scenarios. This was done to provide more detailed token information. Also added an absolute value conversion in flashloans.ts to prevent cases of negative values.

jakubswierczek committed:
Improve Aave operation functionalities

Improved flashloan handling in Aave operations, and added debtToken symbol for better token information. Introduced a condition to withdraw maximum uint value in case collateral matches flashloan token. This resolves potential issues where negative values could be sent to the smart contract. Furthermore, removed unnecessary console info logging and cleaned up import sequences for better readability.

jakubswierczek committed:
lint fix

robercano commented on packages/domain/src/flashloans.ts:
```diff
@@ -39,6 +39,7 @@ export const transientCollateralFlashloan: TransientCollateralFlashloan = (
   )
     .times(oraclePrice)
     .div(maxLoanToValueWhenCollateralising.times(ONE.minus(FLASHLOAN_SAFETY_MARGIN)))
+    .abs()
```
Why is this `.abs()` needed?

robercano commented on packages/dma-library/src/strategies/aave/common/get-flashloan-token.ts:
```diff
@@ -1,15 +1,32 @@
+  protocol,
+  debt,
+}: FlashloanDependencies): WithFlashLoanArgs {
+  if (protocol === 'Spark') {
```
Is it possible to use an `enum` here instead of a harcoded string?

jakubswierczek replied:
The protocol variable is of type: `export type AaveLikeProtocol = 'AAVE' | 'AAVE_V3' | 'Spark'`, so it's not hardcoded per se. ;)

robercano commented on packages/dma-library/src/strategies/aave-like/multiply/common/build-flashloan-sim-args.ts:
```diff
@@ -1,23 +1,29 @@
  */
-export function buildFlashloanSimArgs(
-  flashloanTokenAddress: string,
+export async function buildFlashloanSimArgs(
```
Why is this `async`? I don't see any calls that need `await`

jakubswierczek replied:
My apologies, I forgot to remove it after conducting my tests.

robercano commented on packages/dma-library/src/strategies/aave-like/multiply/close/types.ts:
```diff
@@ -1,16 +1,28 @@
+  } & WithFlashLoanArgs
+
+export type CloseFlashloanArgs = {
+  token: {
```
I think we could create a `Token` type being:
```
export type Token : {
  address: string
  precision: number
  symbol: string
}
```

and then a `TokenAmount` type being:
```
export type TokenAmount:  {
  token: Token
  amount: BigNumber
}
```

To try to start having a bit more semantics in the code

jakubswierczek replied:
Yes it's a good idea!

robercano reviewed (changes requested):
Looks good in general! I'm missing the related tests, though. Is this done somewhere else?

jakubswierczek committed:
Refactor flashloan simulation to synchronous

This commit removes the unnecessary asynchronous handling in the buildFlashloanSimArgs function and associated calls in the 'simulate.ts' files of both open and adjust strategies. The async handler was superfluous as there were no asynchronous operations within the function, thus simplifying complexity and improving performance.

robercano reviewed (approved):
Agreed to make the following changes in a future PR instead of in this one:
- Use `Protocol.SPARK` from `dma-library/src/types` instead of hardcoded string
- Review the use of `.abs()` when calculating flashloans to see if it is still needed
- Agreed to use `Token` and `TokenAmount` types

jakubswierczek committed:
v0.5.9-rc-flashloan.1

jakubswierczek committed:
Merge branch 'dev' into jacob/sc-11917/flashloan-information

# Conflicts:
#	packages/dma-library/package.json

jakubswierczek committed:
v0.5.10-rc-flashloan.1

jakubswierczek committed:
v0.5.10
